### PR TITLE
sinit: build without depending on ubase

### LINKED ISSUE #273: sinit: dependency on ubase requires kiss alternatives intervention
Currently sinit requires respawn (currently of the ubase package) to keep your getty alive (busybox init uses inittab to accomplish this).

While we were at it, [building the sinit package], to kill two birds with one stone, the included sinit scripts also default to using ubase-box getty. I consider this a nice touch.

The problem is that this was all devised before the new alternatives system became available. We have since upgraded ubase-box to take advantage of alternatives, which however requires manual intervention to install.

How do we address this?

### PR DESCRIPTION
**Pull request by cemkeylan**
Resolves #273 

Okay, so this first commit might be a little rough on the edges. I have added a modified version of the build as packaged in Carbs. I double checked for different directory names (`/etc/init` to `/etc/rc.d`) but I might have something mixed.

I will talk about the rationale of all the changes.

### Adding ubase halt and respawn

This is obviously a given if we don't want to add ubase dependency just for two simple utilities. `busybox halt` is not compatible with `sinit`. Also, sinit doesn't have a built-in respawn function, we probably require `respawn` from ubase.

I am saying "probably" here, because we don't necessarily depend on it. You don't have to restart getty if it exits, but realistically speaking, you would want getty to restart if you exit it. Therefore, it is built-in to the package.

### Adding poweroff and reboot scripts

Replacing init should feel natural. Every init system comes with a complementary `poweroff` and `reboot` tools. Nobody would actually want to write `/bin/kill -s USR1 1` every time they want to shut their computers down.

### Boot and shutdown hooks

I have removed the `pre.shutdown` hook and changed the `post.shutdown`.

```sh
# We handle the shutdown from here

if [ "$SINIT_SHUTDOWN_HOOKS" = 1 ]; then
        case "$1" in
                reboot)
                        out "Requesting system reboot..."
                        halt -r
                        ;;
                *)     # I have changed this from shutdown to *
                        out "Requesting system poweroff..."
                        halt -p
                        ;;
        esac
fi
```
Now that our sinit build is quite stable, we don't need a pre-shutdown check to control the validity of the variable. Just shutdown.

I have also merged the boot-hook onto a single file to avoid crowd.

### Issues

* ~~I have a leftover `# vim:filetype=sh` on `post.shutdown`~~

### CODE REVIEW

File: community/sinit/files/poweroff [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
```
Comment (dylanaraps) [inline community/sinit/files/poweroff]: Remove this newline.

Comment (cemkeylan) [reply community/sinit/files/poweroff]: :+1:

File: community/sinit/build [hunk 2]
```diff
@@ -1,10 +1,22 @@
 #!/bin/sh -e
 
-make CFLAGS="$CFLAGS -static"
+mv config.h sinit/config.h
 
-install -Dm 755 sinit "$1/usr/bin/sinit"
+cd sinit
+make CFLAGS="$CFLAGS -static" PREFIX=/usr DESTDIR="$1" install
+ln -s /usr/bin/sinit "$1/usr/bin/init"
```
Comment (dylanaraps) [inline community/sinit/build]: Can wrap this in `()` to avoid `cd ..`.

Comment (cemkeylan) [reply community/sinit/build]: In a function like below?

```sh
buildsinit() {
    cd sinit
    make (...) install
    ln -s (...)
}

**Comment by a-schaefers** [comment]
Can we explore shell methods of replacing ubase requirement of sinit?

instead of `respawn`,

`while :; do /sbin/getty 34800 "tty${getty}" &>/dev/null; done &`

instead of `halt`,

https://major.io/2009/01/29/linux-emergency-reboot-or-shutdown-with-magic-commands/

```
echo 1 > /proc/sys/kernel/sysrq 
echo o > /proc/sysrq-trigger
```

**Comment by cemkeylan** [comment]
It seems like a good alternative, but since halt and respawn are relatively small (around 32K in total) I would prefer keeping them. They only have a single purpose and they are good at them. POSIX shell is not a monstrosity either but we have small tools for the same purpose, making them more reliable, fast, and secure. 

I am also not so quite sure about the safety of halting the computer from sysrq, but I might be wrong.

**Comment by a-schaefers** [comment]
Since we already use posix sh killall5 and baseinit it seems like a really good fit to go shell all the way here. I will be a little sad if we don't try this. Re: safety, it should be entirely safe to the best of my knowledge. But I will leave the final decision to you and @dylanaraps . Thanks.

**Comment by cemkeylan** [comment]
I have tested them both, I can safely say that they are working. I could replace them on the community repository, but I wouldn't necessarily depend on them for my own buid. We do need SYSRQ enabled in the kernel config, but most of the people have it enabled. It's on the default configuration.

File: community/sinit/build (hunk 2 above)
Comment (dylanaraps) [reply community/sinit/build]: No no,

```sh
(
    cd sinit
    make bla bla
    ln -s bla la
)
```

Then you can avoid `cd ..` as the `cd` is kept local to the braces.

**Comment by cemkeylan** [comment]
Will do 👍

On 30 January 2020 13:33:53 GMT+03:00, dylan <notifications@github.com> wrote:
>dylanaraps commented on this pull request.
>
>
>
>> +cd sinit
>+make CFLAGS="$CFLAGS -static" PREFIX=/usr DESTDIR="$1" install
>+ln -s /usr/bin/sinit "$1/usr/bin/init"
>
>No no,
>
>```sh
>(
>    cd sinit
>    make bla bla
>    ln -s bla la
>)
>```
>
>Then you can avoid `cd ..` as the `cd` is kept local to the braces.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/kisslinux/community/pull/276#discussion_r372871990

-- 
Cem Keylan

File: community/sinit/post-install [hunk 3]
```diff
@@ -1,33 +1,35 @@
 #!/bin/sh -e
 
-cat <<EOF
-
-IMPORTANT:
-        This package is installed as /usr/bin/sinit
-        Please update your kernel parameters
-        accordingly,
+grep -q SINIT_ /etc/rc.conf || cat <<EOF >> /etc/rc.conf
```
[the remark this answers is not in the document]
Comment (cemkeylan) [reply community/sinit/post-install]: I didn't know you have talked about this, but I really don't think that this is an issue. We could just inform the user that we have added these comments to their rc.conf. Since these are just comments, the only impact this has to the user's system is that it saves the user the time to copy and paste it on their own.

File: community/sinit/files/sinit-launch-services.boot [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+sinit_run_getty() {
+    for getty in 1 2 3 4 5 6; do
+        respawn /sbin/getty 38400 tty${getty} 2>&1 &
+    done
+}
+
+# for ubase
+# sinit_run_getty() {
+#     for getty in 1 2 3 4 5 6 ; do
+#         respawn /sbin/getty /dev/tty${getty} linux 2>&1 &
```
[the remark this answers is not in the document]
Comment (cemkeylan) [reply community/sinit/files/sinit-launch-services.boot]: Yeah, I'll do that.

File: community/sinit/post-install (hunk 3 above)
[the remark this answers is not in the document]
Comment (a-schaefers) [reply community/sinit/post-install]: TBH I can't remember exactly how it all happened. The change is fine with me, personally. I'm just not sure it's a best practice however to just append things to peoples config files in /etc/

[the remark this answers is not in the document]
Comment (cemkeylan) [reply community/sinit/post-install]: Yeah, it is not the best practice, but since it is only a comment, it can be justified. 

We could also modify init and make it read every `*.conf` file on `/etc/rc.d`, the way we do it for hooks. That way we could simply drop an `sinit.conf` to `/etc/rc.d` without directly editing the user's personal configurations.

File: community/sinit/files/sinit-launch-services.boot [hunk 5]
```diff
@@ -3,18 +3,13 @@
 # for busybox
 sinit_run_getty() {
     for getty in 1 2 3 4 5 6; do
+	# for busybox
         respawn /sbin/getty 38400 tty${getty} 2>&1 &
+	# for ubase
+	# respawn /sbin/getty /dev/tty${getty} linux 2>&1 &
```
Comment (a-schaefers) [inline community/sinit/files/sinit-launch-services.boot]: Personally I prefer # comment after the fork '&' in this sittuation.

```
respawn /sbin/getty 38400 tty${getty} 2>&1 &         # busybox getty
￼# respawn /sbin/getty /dev/tty${getty} linux 2>&1 & # ubase getty
```

But @dylanaraps is the king of nitpicks so I wonder what he thinks?

@cemkeylan Can we use a different word than 'for' in  `# for busybox`

Because to me, all of the uses of the word "for" around the for loop makes it look confusing ;)

Also line 3 can be removed.

Thanks @cemkeylan

File: community/sinit/files/sinit-launch-services.boot [hunk 6]
```diff
@@ -1,6 +1,5 @@
 # we are going to start runit and getty here
 
-# for busybox
```
Comment (cemkeylan) [inline community/sinit/files/sinit-launch-services.boot]: I meant duplicate comment not command

File: community/sinit/build (hunk 2 above)
Comment (dylanaraps) [reply community/sinit/build]: Please indent the code like my example above.

File: community/sinit/build [hunk 7]
```diff
@@ -1,10 +1,25 @@
...
+install -Dm644 sinit-launch-services.boot "$1/etc/rc.d/sinit-launch-services.boot"
+install -Dm644 sinit.post.shutdown "$1/etc/rc.d/sinit.post.shutdown"
+
+(
+cd ubase
+make halt
+make respawn
+
+install -Dm755 halt "$1/usr/bin/halt"
+install -Dm755 respawn "$1/usr/bin/respawn"
```
Comment (dylanaraps) [inline community/sinit/build]: Indent this please.

File: community/sinit/files/sinit.post.shutdown [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+		reboot)	
+			out "Requesting system reboot..."
+			halt -r 
+			;;
+		*) 
+			out "Requesting system poweroff..."
+			halt -p 
+			;;
+	esac
+fi
```
Comment (dylanaraps) [inline community/sinit/files/sinit.post.shutdown]: 4 spaces for indentation.

**Comment by dylanaraps** [comment]
Do we really need to use `respawn`? Can we not just use an infinite loop or recursion?

**Comment by cemkeylan** [comment]
:+1:

File: community/sinit/build [hunk 9]
```diff
@@ -1,10 +1,23 @@
...
+
+install -Dm644 sinit-launch-services.boot "$1/etc/rc.d/sinit-launch-services.boot"
+install -Dm644 sinit.post.shutdown "$1/etc/rc.d/sinit.post.shutdown"
+
+(
+    cd ubase
+    make halt
+
+    install -Dm755 halt "$1/usr/bin/halt"
+)
```
Comment (dylanaraps) [inline community/sinit/build]: Is halt needed? Our `rc.shutdown` script doesn't use it at all.

File: community/sinit/files/sinit-launch-services.boot [hunk 10]
```diff
@@ -0,0 +1,15 @@
+# we are going to start runit and getty here
+
+sinit_run_getty() {
+    for getty in 1 2 3 4 5 6; do
+        while :; do /sbin/getty 38400 tty${getty} 2>&1 ; done &  # busybox getty
+        # while :; do /sbin/getty /dev/tty${getty} linux 2>&1 ; done &  # ubase getty
```
Comment (dylanaraps) [inline community/sinit/files/sinit-launch-services.boot]: Is there not a standard set of `getty` arguments which work regardless of busybox/ubase/etc getty?

File: community/sinit/files/sinit.post.shutdown [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+        reboot)	
+            out "Requesting system reboot..."
+            halt -r 
+            ;;
+        *) 
+            out "Requesting system poweroff..."
+            halt -p 
+            ;;
+    esac
+fi
```
Comment (dylanaraps) [inline community/sinit/files/sinit.post.shutdown]: Why is this needed?

File: community/sinit/files/sinit-launch-services.boot (hunk 10 above)
Comment (cemkeylan) [reply community/sinit/files/sinit-launch-services.boot]: Nope, there is no standard regarding getty arguments.

File: community/sinit/build (hunk 9 above)
Comment (cemkeylan) [reply community/sinit/build]: sinit.post.shutdown requires it.

File: community/sinit/files/sinit.post.shutdown (hunk 11 above)
Comment (cemkeylan) [reply community/sinit/files/sinit.post.shutdown]: If we don't do that system will not poweroff and we will be stranded without a getty and system services on an empty screen.

File: community/sinit/build (hunk 9 above)
Comment (cemkeylan) [reply community/sinit/build]: It could be replaced with @a-schaefers 's idea of shutting down from /proc. But I don't know how safe that is.

File: community/sinit/files/sinit-launch-services.boot (hunk 10 above)
Comment (cemkeylan) [reply community/sinit/files/sinit-launch-services.boot]: They are really incompatible, with ubase you must

* NOT specify BAUD_RATE
* Specify full path of tty
* Specify terminal type (linux)

With busybox you must

* Specify BAUD_RATE
* NOT specify full path
* Terminal type is optional

Comment (dylanaraps) [reply community/sinit/files/sinit-launch-services.boot]: This *should* work for both: `getty /dev/tty3 0 linux` (example).

File: community/sinit/files/sinit.post.shutdown (hunk 11 above)
Comment (dylanaraps) [reply community/sinit/files/sinit.post.shutdown]: How does busybox init work around this? Does the `shutdown` command not trigger this for us?

**Comment by dylanaraps** [comment]
I'd like to find some middle ground between duplicating everything for other init systems if at all possible. :+1:

File: community/sinit/files/sinit.post.shutdown (hunk 11 above)
Comment (cemkeylan) [reply community/sinit/files/sinit.post.shutdown]: As far as i know, it is built inside init. System halts when the shutdown functions on the inittab are complete.

**Comment by cemkeylan** [comment]
We could theoretically build a halt function inside sinit, but that would defeat the whole purpose of 'simple init'. In terms of compatability with other inits, the only thing compatible with busybox init is sysvinit. We can't possibly provide this compatability without altering some functionality.

File: community/sinit/files/poweroff [hunk 12]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+/bin/kill -s USR1 1
```
Comment (dylanaraps) [inline community/sinit/files/poweroff]: Any reason why we can't drop `/bin/`?

File: community/sinit/files/reboot [hunk 13]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+/bin/kill -s INT 1
```
Comment (dylanaraps) [inline community/sinit/files/reboot]: Any reason why we can't drop `/bin/`?

File: community/sinit/post-install (hunk 3 above)
[the remark this answers is not in the document]
Comment (dylanaraps) [reply community/sinit/post-install]: I'm against modifying user config files like this. I'd much rather prefer a post-install message. :+1:

File: community/sinit/post-install [hunk 14]
```diff
@@ -1,33 +1,35 @@
...
-        -> SINIT_ENABLE_GETTY=1            # <- (required)
-        -> SINIT_ENABLE_RUNIT=1            # <- (optional)
+	Init is installed to /usr/bin/sinit, you 
+	can also link it to /usr/bin/init but you 
+	should only do that on a chroot environment.
 
-USAGE:
-        To shutdown:
+	To completely replace busybox init programs,
+	you can do (while chrooted)
```
Comment (dylanaraps) [inline community/sinit/post-install]: > while chrooted

This can be removed?

File: community/sinit/post-install [hunk 15]
```diff
@@ -1,33 +1,35 @@
...
+        If you want to start using sinit, you 
+	need to edit your /etc/rc.conf to enable
+	SINIT variables. 
 
-        -> SINIT_ENABLE_SHUTDOWN_HOOKS=1   # <- (required)
-        -> SINIT_ENABLE_GETTY=1            # <- (required)
-        -> SINIT_ENABLE_RUNIT=1            # <- (optional)
+	Init is installed to /usr/bin/sinit, you 
+	can also link it to /usr/bin/init but you 
+	should only do that on a chroot environment.
```
Comment (dylanaraps) [inline community/sinit/post-install]: > should only do that on a chroot environment.

What do you mean by this?

File: community/sinit/sources [hunk 16]
```diff
@@ -1,6 +1,7 @@
-https://dl.suckless.org/sinit/sinit-1.0.tar.gz
-files/config.def.h
-files/etc/rc.d/sinit.getty.boot
-files/etc/rc.d/sinit.runit.boot
-files/etc/rc.d/sinit.pre.shutdown
-files/etc/rc.d/sinit.post.shutdown
+https://dl.suckless.org/sinit/sinit-1.0.tar.gz sinit
+https://github.com/michaelforney/ubase/archive/b33bdaa.tar.gz ubase
+files/config.h
```
Comment (dylanaraps) [inline community/sinit/sources]:
```sh
files/config.h sinit
```

You can then remove the `mv` for this in `build`.

File: community/sinit/build (hunk 9 above)
Comment (dylanaraps) [reply community/sinit/build]: Shutting down from `/proc` requires SYSRQ to be enabled in your kernel. Not always going to be the case (isn't in my config).

File: community/sinit/files/sinit.post.shutdown [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+    case "$1" in
+        reboot)	
+            out "Requesting system reboot..."
+            halt -r 
+            ;;
+        *) 
+            out "Requesting system poweroff..."
+            halt -p 
+            ;;
+    esac
```
Comment (dylanaraps) [inline community/sinit/files/sinit.post.shutdown]:
```sh
case $1 in
    reboot)	halt -r ;;
    *)      halt -p ;;
esac
```

**Comment by cemkeylan** [comment]
Because kill is not standard in shells. If someone calls kill from ash, it is not a builtin, so it will call /bin/kill. However if someone uses bash in POSIX mode linked to /bin/sh, there is a kill builtin that does not support signaling (while specifying signal name is the POSIX thing to do), has different options etc. The /bin is there for ensuring compatability.

On 4 February 2020 16:25:19 GMT+03:00, dylan <notifications@github.com> wrote:
>dylanaraps commented on this pull request.
>
>
>
>> @@ -0,0 +1,2 @@
>+#!/bin/sh
>+/bin/kill -s INT 1
>
>Any reason why we can't drop `/bin/`?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/kisslinux/community/pull/276#pullrequestreview-352970400

-- 
Cem Keylan

**Comment by cemkeylan** [comment]
You have to do a hardware shutdown if you replace init while the other init is still active.

On 4 February 2020 16:27:03 GMT+03:00, dylan <notifications@github.com> wrote:
>dylanaraps commented on this pull request.
>
>
>
>>  
>-        -> SINIT_ENABLE_SHUTDOWN_HOOKS=1   # <- (required)
>-        -> SINIT_ENABLE_GETTY=1            # <- (required)
>-        -> SINIT_ENABLE_RUNIT=1            # <- (optional)
>+	Init is installed to /usr/bin/sinit, you 
>+	can also link it to /usr/bin/init but you 
>+	should only do that on a chroot environment.
>
>> should only do that on a chroot environment.
>
>What do you mean by this?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/kisslinux/community/pull/276#pullrequestreview-352971554

-- 
Cem Keylan

File: community/sinit/build (hunk 9 above)
Comment (a-schaefers) [reply community/sinit/build]: > Shutting down from `/proc` requires SYSRQ to be enabled in your kernel. Not always going to be the case (isn't in my config).

I do not believe this is correct,  echo "1" > /proc/sys/kernel/sysrq should enable SYSRQ features temporarily until the next reboot (in this case, the change is immediately lost and would be harmless)

**Comment by cemkeylan** [comment]
No, that's not true. You don't have a /proc/sys/kernel/sysrq if your kernel is not configured like that.

On 4 February 2020 21:35:01 GMT+03:00, Adam Schaefers <notifications@github.com> wrote:
>a-schaefers commented on this pull request.
>
>
>
>> +(
>+    cd ubase
>+    make halt
>+
>+    install -Dm755 halt "$1/usr/bin/halt"
>+)
>
>> Shutting down from `/proc` requires SYSRQ to be enabled in your
>kernel. Not always going to be the case (isn't in my config).
>
>I do not believe this is correct,  echo "1" > /proc/sys/kernel/sysrq
>should enable SYSRQ features temporarily until the next reboot (in this
>case, the change is immediately lost and would be harmless)
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/kisslinux/community/pull/276#discussion_r374848654

-- 
Cem Keylan

**Comment by a-schaefers** [comment]
> You have to do a hardware shutdown if you replace init while the other init is still active.
> […](#)
> On 4 February 2020 16:27:03 GMT+03:00, dylan ***@***.***> wrote: dylanaraps commented on this pull request. > - -> SINIT_ENABLE_SHUTDOWN_HOOKS=1 # <- (required) - -> SINIT_ENABLE_GETTY=1 # <- (required) - -> SINIT_ENABLE_RUNIT=1 # <- (optional) + Init is installed to /usr/bin/sinit, you + can also link it to /usr/bin/init but you + should only do that on a chroot environment. > should only do that on a chroot environment. What do you mean by this? -- You are receiving this because you were mentioned. Reply to this email directly or view it on GitHub: [#276 (review)](https://github.com/kisslinux/community/pull/276#pullrequestreview-352971554)
> -- Cem Keylan

I am not so sure this is always true. We only replaced "/bin/init" on the _filesystem_, but it is running actively in _memory_. The trouble comes when you replace /bin/reboot command. This might be worked around via typing out the full `busybox reboot` command on the first reboot.

**Comment by cemkeylan** [comment]
Yeah well I did it back when sinit was not packaged. Try it out.

On 4 February 2020 21:39:18 GMT+03:00, Adam Schaefers <notifications@github.com> wrote:
>> You have to do a hardware shutdown if you replace init while the
>other init is still active.>
>> […](#)>
>> On 4 February 2020 16:27:03 GMT+03:00, dylan ***@***.***> wrote:
>dylanaraps commented on this pull request. > - ->
>SINIT_ENABLE_SHUTDOWN_HOOKS=1 # <- (required) - -> SINIT_ENABLE_GETTY=1
># <- (required) - -> SINIT_ENABLE_RUNIT=1 # <- (optional) + Init is
>installed to /usr/bin/sinit, you + can also link it to /usr/bin/init
>but you + should only do that on a chroot environment. > should only do
>that on a chroot environment. What do you mean by this? -- You are
>receiving this because you were mentioned. Reply to this email directly
>or view it on GitHub: [#276
>(review)](https://github.com/kisslinux/community/pull/276#pullrequestreview-352971554)>
>> -- Cem Keylan>
>>
>I am not so sure this is always true. We only replaced "/bin/init" on
>the _filesystem_, but it is running actively in _memory_. The trouble
>comes when you replace /bin/reboot command. This might be worked around
>via typing out the full `busybox reboot` command on the first reboot.>
>>
>-- >
>You are receiving this because you were mentioned.>
>Reply to this email directly or view it on GitHub:>
>https://github.com/kisslinux/community/pull/276#issuecomment-582054143

-- 
Cem Keylan

**Comment by a-schaefers** [comment]
> Yeah well I did it back when sinit was not packaged. Try it out.
> […](#)
> On 4 February 2020 21:39:18 GMT+03:00, Adam Schaefers ***@***.***> wrote: > You have to do a hardware shutdown if you replace init while the other init is still active.> > […](#)> > On 4 February 2020 16:27:03 GMT+03:00, dylan ***@***.***> wrote: dylanaraps commented on this pull request. > - -> SINIT_ENABLE_SHUTDOWN_HOOKS=1 # <- (required) - -> SINIT_ENABLE_GETTY=1 # <- (required) - -> SINIT_ENABLE_RUNIT=1 # <- (optional) + Init is installed to /usr/bin/sinit, you + can also link it to /usr/bin/init but you + should only do that on a chroot environment. > should only do that on a chroot environment. What do you mean by this? -- You are receiving this because you were mentioned. Reply to this email directly or view it on GitHub: [#276 (review)]([#276 (review)](https://github.com/kisslinux/community/pull/276#pullrequestreview-352971554))> > -- Cem Keylan> > I am not so sure this is always true. We only replaced "/bin/init" on the _filesystem_, but it is running actively in _memory_. The trouble comes when you replace /bin/reboot command. This might be worked around via typing out the full `busybox reboot` command on the first reboot.> > -- > You are receiving this because you were mentioned.> Reply to this email directly or view it on GitHub:> [#276 (comment)](https://github.com/kisslinux/community/pull/276#issuecomment-582054143)
> -- Cem Keylan

I am not sure what you mean here.

File: community/sinit/build (hunk 9 above)
Comment (dylanaraps) [reply community/sinit/build]: From my system (without SYSRQ):

```sh
# This is as root.
-> cd /proc/sys/kernel/
-> pwd
/proc/sys/kernel
-> echo "1" > /proc/sys/kernel/sysrq
sh: can't create /proc/sys/kernel/sysrq: nonexistent directory
```

Comment (dylanaraps) [reply community/sinit/build]: Also:

```sh
-> ls -ld /proc/sys/kernel/
dr-xr-xr-x    1 root     root             0 Feb  4 15:19 /proc/sys/kernel/
```

**Comment by a-schaefers** [comment]
> No, that's not true. You don't have a /proc/sys/kernel/sysrq if your kernel is not configured like that.
I wonder how obscure a person is who doesn't have SYSRQ enabled or if it is a common decision? If it is rare, one could just put add a line about SYSRQ to postinstall @dylanaraps

**Comment by cemkeylan** [comment]
I mean i did what you told me when I first installed sinit. I don't remember what problems it gave me, but I really don't want to reinstall busybox-init to reproduce it.

On 4 February 2020 21:46:22 GMT+03:00, Adam Schaefers <notifications@github.com> wrote:
>> Yeah well I did it back when sinit was not packaged. Try it out.>
>> […](#)>
>> On 4 February 2020 21:39:18 GMT+03:00, Adam Schaefers ***@***.***>
>wrote: > You have to do a hardware shutdown if you replace init while
>the other init is still active.> > […](#)> > On 4 February 2020
>16:27:03 GMT+03:00, dylan ***@***.***> wrote: dylanaraps commented on
>this pull request. > - -> SINIT_ENABLE_SHUTDOWN_HOOKS=1 # <- (required)
>- -> SINIT_ENABLE_GETTY=1 # <- (required) - -> SINIT_ENABLE_RUNIT=1 #
><- (optional) + Init is installed to /usr/bin/sinit, you + can also
>link it to /usr/bin/init but you + should only do that on a chroot
>environment. > should only do that on a chroot environment. What do you
>mean by this? -- You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub: [#276
>(review)]([#276
>(review)](https://github.com/kisslinux/community/pull/276#pullrequestreview-352971554))>
>> -- Cem Keylan> > I am not so sure this is always true. We only
>replaced "/bin/init" on the _filesystem_, but it is running actively in
>_memory_. The trouble comes when you replace /bin/reboot command. This
>might be worked around via typing out the full `busybox reboot` command
>on the first reboot.> > -- > You are receiving this because you were
>mentioned.> Reply to this email directly or view it on GitHub:> [#276
>(comment)](https://github.com/kisslinux/community/pull/276#issuecomment-582054143)>
>> -- Cem Keylan>
>>
>I am not sure what you mean here.>
>>
>-- >
>You are receiving this because you were mentioned.>
>Reply to this email directly or view it on GitHub:>
>https://github.com/kisslinux/community/pull/276#issuecomment-582057195

-- 
Cem Keylan

**Comment by cemkeylan** [comment]
Why depend on a kernel configuration when you could not depend on a kernel configuration?

On 4 February 2020 21:48:42 GMT+03:00, Adam Schaefers <notifications@github.com> wrote:
>> No, that's not true. You don't have a /proc/sys/kernel/sysrq if your
>kernel is not configured like that.
>I wonder how obscure a person is who doesn't have SYSRQ enabled or if
>it is a common decision? If it is rare, one could just put add a line
>about SYSRQ to postinstall @dylanaraps 
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/kisslinux/community/pull/276#issuecomment-582058170

-- 
Cem Keylan

**Comment by dylanaraps** [comment]
> Why depend on a kernel configuration when you could not depend on a kernel configuration?

You'd be surprised at how much of your system already relies on your kernel configuration. I had a lot of fun starting with a `make allnoconfig` and working up from there boot by boot.

That being said, if it can be avoided, I'd rather avoid it. 

In place of SYSRQ we can use a very simple C program to tell the kernel to shutdown and ship it with `baseinit`. I'll work on this shortly.

**Comment by cemkeylan** [comment]
> In place of SYSRQ we can use a very simple C program to tell the kernel to shutdown [...]

We could add ubase halt with a different name to baseinit.

**Comment by dylanaraps** [comment]
It's far simpler to not pull in code from `ubase`. See: https://github.com/kisslinux/init/blob/master/bin/kpow.c

**Comment by cemkeylan** [comment]
:+1:

**Comment by dylanaraps** [comment]
Added code to `rc.shutdown` to run `kpow` if needed. This won't conflict with busybox init and should also be compatible with other inits down the line. The `kpow` stuff can be removed from this PR. :+1:

https://github.com/kisslinux/init/commit/b095bfb1509c14d951bdcf78049e745394fd95aa

**Comment by cemkeylan** [comment]
:+1:

**Comment by cemkeylan** [comment]
Removing halt now.

File: community/sinit/build [hunk 18]
```diff
@@ -1,10 +1,9 @@
 #!/bin/sh -e
 
-make CFLAGS="$CFLAGS -static"
+make CFLAGS="$CFLAGS -static" PREFIX=/usr DESTDIR="$1" install
+ln -s /usr/bin/sinit "$1/usr/bin/init"
```
Comment (dylanaraps) [inline community/sinit/build]:
```sh
ln -s sinit "$1/usr/bin/init"
```

Comment (cemkeylan) [reply community/sinit/build]: :+1:

**Comment by dylanaraps** [comment]
This should cover the transition period from a running machine. :+1:

```sh
#!/bin/sh -e

cat <<EOF

IMPORTANT:

    To start using sinit, the following variables
    must be added to your /etc/rc.conf.

	    [1mSINIT_ENABLE_GETTY=1[m   (Mandatory)
	    [1mSINIT_ENABLE_RUNIT=1[m   (Optional)

    sinit is installed as /usr/bin/sinit. It can fully
    replace busybox init by running the following command.

	    kiss a | grep ^sinit | kiss a -

    NOTE: If swapping from busybox init on a running system,
          shutdown/reboot by invoking 'busybox shutdown'.
          sinit's poweroff/shutdown utilities can't be of
          use until it is actually running.

EOF
```

**Comment by cemkeylan** [comment]
Here is the issue, invoking `busybox shutdown` will not work either if you replace /usr/bin/init on a running system.

**Comment by dylanaraps** [comment]
It should work regardless as the `poweroff`/`reboot` commands simply send a signal to the running `init` process no? I'll test on my machine.

**Comment by cemkeylan** [comment]
Yeah that maybe because of my setup. I have a seperate package for busybox-init.

On 4 February 2020 23:32:53 GMT+03:00, dylan <notifications@github.com> wrote:
>It should work regardless as the `poweroff`/`reboot` commands simply
>send a signal to the running `init` process no? I'll test on my
>machine.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/kisslinux/community/pull/276#issuecomment-582101795

-- 
Cem Keylan

**Comment by dylanaraps** [comment]
It worked just fine. :+1:

File: community/sinit/build (hunk 9 above)
Comment (a-schaefers) [reply community/sinit/build]: bummer

**Comment by dylanaraps** [comment]
Any updates on this? 

I'm happy to take over this package if you like. :+1:

**Comment by cemkeylan** [comment]
I will clean everything up and push the final (hopefully) patch tonight. Sorry for the delay.

On 6 February 2020 18:38:26 GMT+03:00, dylan <notifications@github.com> wrote:
>Any updates on this? 
>
>I'm happy to take over this package if you like. :+1:
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/kisslinux/community/pull/276#issuecomment-582963694

-- 
Cem Keylan

**Comment by dylanaraps** [comment]
All good. :+1:

File: community/sinit/files/sinit-launch-services.boot (hunk 10 above)
Comment (cemkeylan) [reply community/sinit/files/sinit-launch-services.boot]: This does not work for ubase, by the way.

**Comment by cemkeylan** [comment]
Again, sorry for the delayed update. 

I thought it would make more sense to simply comment out the functions we are supposed to run during the boot hook stage. We don't need to check some variable, the user can simply uncomment these to use them.

File: community/sinit/sources (hunk 16 above)
Comment (dylanaraps) [reply community/sinit/sources]: This may need fixing again.

File: community/sinit/files/sinit-launch-services.boot [hunk 19]
```diff
@@ -0,0 +1,12 @@
+# We are going to start runit and getty here
+
+# Uncomment below to run getty
+
```
Comment (dylanaraps) [inline community/sinit/files/sinit-launch-services.boot]:
```sh
# Uncomment to enable gettys.
```

(Without extra newline)

File: community/sinit/files/sinit-launch-services.boot [hunk 20]
```diff
@@ -0,0 +1,12 @@
...
+# for getty in 1 2 3 4 5 6 ; do
+#     while :; do /sbin/getty tty${getty} 0 linux ; done &  # busybox getty
+# #    while :; do /sbin/getty tty${getty} linux ; done &   # ubase getty
+# done
+
+# Uncomment below to enable runit
+
```
Comment (dylanaraps) [inline community/sinit/files/sinit-launch-services.boot]:
```sh
# Uncomment to enable runit services
```

(Without extra newline)

File: community/sinit/post-install [hunk 21]
```diff
@@ -2,32 +2,25 @@
...
-
-        -> # kill -s USR1 1
-
-        To reboot:
-
-        -> # kill -s INT 1
+!! IMPORTANT !!
+
+If you want to start using sinit, you 
+need to edit
```
Comment (dylanaraps) [inline community/sinit/post-install]: > To start using sinit, edit

**Comment by cemkeylan** [comment]
:+1:
